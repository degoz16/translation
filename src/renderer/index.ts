import MarkdownIt, {Options} from 'markdown-it';
import Renderer from 'markdown-it/lib/renderer';
import Token from 'markdown-it/lib/token';

export enum CustomRendererLifeCycle {
    BeforeRender,
    AfterRender,
    BeforeInlineRender,
    AfterInlineRender,
}

function isCustomRendererLifeCycle(cycle: unknown): cycle is CustomRendererLifeCycle {
    return Object.keys(CustomRendererLifeCycle).includes(
        cycle as keyof typeof CustomRendererLifeCycle,
    );
}

export type CustomRendererParams<S = {}> = {
    hooks?: CustomRendererHooks;
    rules?: Renderer.RenderRuleRecord;
    state?: S;
};

export type CustomRendererHooks = Record<string, CustomRendererHook | CustomRendererHook[]>;

export interface CustomRendererHook {
    (parameters: CustomRendererHookParameters): string;
}

export type CustomRendererHookParameters = {
    tokens: Token[];
    options: Options;
    env: unknown;
    map?: [number, number] | null;
    // accumulates render artifacts inside the inline render
    rendered?: string[];
};

type SourceMap = [number, number];

function getMap(tokens: Token[], i: number) {
    let depth = 0;
    while (tokens[i]) {
        const token = tokens[i--];

        depth += token.nesting;

        if (token.map && depth >= 0) {
            return token.map;
        }
    }

    return null;
}

export class CustomRenderer<State = {}, Md extends MarkdownIt = MarkdownIt> extends Renderer {
    readonly state: State;

    readonly md: Md;

    protected hooks: Map<CustomRendererLifeCycle, CustomRendererHook[]>;

    constructor(
        {hooks = {}, rules = {}, state = {} as State}: CustomRendererParams<State>,
        parser: Md,
    ) {
        super();

        this.setRules(rules);

        this.md = parser;
        this.state = state;

        this.hooks = new Map<CustomRendererLifeCycle, CustomRendererHook[]>();
        this.setHooks({...hooks});
    }

    setRules(rules: Renderer.RenderRuleRecord) {
        for (const [name, rule] of Object.entries(rules)) {
            if (!rule || !name?.length) {
                continue;
            }

            this.rules[name] = rule.bind(this);
        }
    }

    setHooks(hooks: CustomRendererHooks) {
        for (const [name, hook] of Object.entries(hooks)) {
            if (isCustomRendererLifeCycle(name)) {
                this.hook(parseInt(name, 10), hook);
            }
        }
    }

    hook(cycle: CustomRendererLifeCycle, hook: CustomRendererHook | CustomRendererHook[]) {
        const hooks = this.hooks.get(cycle) ?? [];

        const normalized = (Array.isArray(hook) ? hook : [hook]).map((h) => h.bind(this));

        this.hooks.set(cycle, [...hooks, ...normalized]);
    }

    render(tokens: Token[], options: Options, env: unknown) {
        let rendered = '';

        let children;
        let type;
        let map;
        let len;
        let i;

        const parameters = {tokens, options, env};

        rendered += this.runHooks(CustomRendererLifeCycle.BeforeRender, parameters);

        for (i = 0, len = tokens.length; i < len; i++) {
            type = tokens[i].type;
            children = tokens[i].children;
            map = getMap(tokens, i - 1);

            if (type === 'inline' && Array.isArray(children)) {
                rendered += this.renderInline(children, options, env, map, tokens[i]);

                continue;
            }

            rendered += this.processToken(tokens, i, options, env);
        }

        rendered += this.runHooks(CustomRendererLifeCycle.AfterRender, parameters);

        return rendered;
    }

    // @ts-ignore
    renderInline(
        tokens: Token[],
        options: Options,
        env: unknown,
        map: SourceMap | null,
        inline: Token,
    ) {
        const rendered: string[] = [];
        const parameters = {tokens, options, env, rendered, map, inline};

        rendered.push(this.runHooks(CustomRendererLifeCycle.BeforeInlineRender, parameters));

        for (let i = 0; i < tokens.length; i++) {
            rendered.push(this.processToken(tokens, i, options, env));
        }

        rendered.push(this.runHooks(CustomRendererLifeCycle.AfterInlineRender, parameters));

        return rendered.join('');
    }

    // renderInline provides rendered array
    // we accumulate render results into it
    // allowing us to access current render artifacts
    // at the time each before/after inline render hook runs
    runHooks(cycle: CustomRendererLifeCycle, parameters: CustomRendererHookParameters) {
        const hooks = this.hooks.get(cycle) ?? [];

        let rendered = '';
        let result = '';

        for (const hook of hooks) {
            result = hook(parameters);

            if (Array.isArray(parameters.rendered)) {
                parameters.rendered.push(result);
            } else {
                rendered += result;
            }
        }

        return Array.isArray(parameters.rendered) ? '' : rendered;
    }

    processToken(tokens: Token[], i: number, options: Options, env: unknown) {
        let rendered = '';

        const type = tokens[i].type;
        const rule = this.rules[type];

        if (rule) {
            // @ts-ignore
            rendered += rule(tokens, i, options, env, this);
        } else {
            rendered += this.renderToken(tokens, i, options);
        }

        return rendered;
    }
}

export function customRenderer<State = {}, Md extends MarkdownIt = MarkdownIt>(
    parser: Md,
    parameters: CustomRendererParams<State> = {},
) {
    // @ts-ignore
    // eslint-disable-next-line no-param-reassign
    parser.renderer = new CustomRenderer<State, Md>(parameters, parser);
}
