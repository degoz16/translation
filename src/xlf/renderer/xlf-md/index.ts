import assert from 'assert';
import {XLFTagToken, XLFTextToken, XLFToken, isXLFTagToken, isXLFTextToken} from 'src/xlf/token';

export type XLFMDRendererRuleSet = {
    text: XLFMDRendererRule;
    [key: string]: XLFMDRendererRule;
};

export type XLFMDRendererRule = (token: XLFToken) => string;

const literal = (token: XLFToken): string => {
    assert(isXLFTagToken(token));
    token as XLFTagToken;

    const {equivText} = token;
    assert(
        equivText?.length,
        'literal tag should contain original markup inside equiv-text attritbute',
    );

    return equivText;
};

const spaced = (actor: (token: XLFToken) => string) => (token: XLFToken) => ' ' + actor(token);
const quoted = (actor: (token: XLFToken) => string) => (token: XLFToken) =>
    '"' + actor(token) + '"';
const attr = spaced(quoted(literal));

class XLFMDRenderer {
    rules: XLFMDRendererRuleSet;

    constructor() {
        this.rules = {
            text: this.text.bind(this),
            tag: this.tag.bind(this),
            x: this.x.bind(this),

            // simple pair syntax
            // strong
            strong_open: literal,
            strong_close: literal,
            // em
            em_open: literal,
            em_close: literal,
            // s
            s_open: literal,
            s_close: literal,
            // sup
            sup_open: literal,
            sup_close: literal,
            // samp
            samp_open: literal,
            samp_close: literal,
            // code
            code_open: literal,
            code_close: literal,

            // link
            link_text_part_open: literal,
            link_text_part_close: literal,
            link_attributes_part_open: literal,
            link_attributes_part_close: literal,
            link_attributes_href: literal,
            link_attributes_title: attr,
            link_autolink: literal,
            link_reflink: literal,

            // image
            image_text_part_open: literal,
            image_text_part_close: literal,
            image_attributes_part_open: literal,
            image_attributes_part_close: literal,
            image_attributes_src: literal,
            image_attributes_title: attr,
            image_attributes_size: spaced(literal),

            // video
            video: literal,

            // anchor
            anchor: literal,

            // file
            file_open: literal,
            file_src: spaced(literal),
            file_name_open: spaced(literal),
            file_name_close: literal,
            file_referrerpolicy: spaced(literal),
            file_rel: spaced(literal),
            file_target: spaced(literal),
            file_type: spaced(literal),
            file_close: spaced(literal),

            // liquid
            liquid_Include: literal,
            liquid_ListTabs: literal,
            liquid_EndListTabs: literal,
            liquid_If: literal,
            liquid_Else: literal,
            liquid_EndIf: literal,
            liquid_ForInLoop: literal,
            liquid_EndForInLoop: literal,
            liquid_Function: literal,
            liquid_Filter: literal,
            liquid_Variable: literal,
            liquid_Attributes: literal,

            // html
            html_inline: literal,
        };
    }

    render(tokens: XLFToken[]): string {
        let rendered = '';

        for (const token of tokens) {
            const handler = this.rules[token.type];
            if (handler) {
                rendered += handler(token);
            }
        }

        return rendered;
    }

    text(token: XLFToken): string {
        assert(isXLFTextToken(token));
        token as XLFTextToken;

        return token.data;
    }

    tag(token: XLFToken): string {
        const handler = this.rules[token.data];
        if (!handler) {
            return '';
        }

        return handler(token);
    }

    x(token: XLFToken): string {
        assert(isXLFTagToken(token));
        token as XLFTagToken;

        const syntax = token.syntax;
        if (!syntax?.length) {
            throw new Error("can't render g tag without syntax");
        }

        const handler = this.rules[syntax];
        if (!handler) {
            throw new Error(`syntax ${syntax} not implemented`);
        }

        return handler(token);
    }
}

export {XLFMDRenderer};
export default {XLFMDRenderer};
