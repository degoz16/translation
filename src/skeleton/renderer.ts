import MarkdownIt from 'markdown-it';
import {
    MarkdownRenderer,
    MarkdownRendererEnv,
    MarkdownRendererParams,
    mdRenderer,
} from '@diplodoc/markdown-it-markdown-renderer';

// configure with diplodoc plugins
// @ts-ignore
import meta from 'markdown-it-meta';
// @ts-ignore
import sup from 'markdown-it-sup';
import notes from '@diplodoc/transform/lib/plugins/notes';
import cut from '@diplodoc/transform/lib/plugins/cut';
import checkbox from '@diplodoc/transform/lib/plugins/checkbox';
import anchors from '@diplodoc/transform/lib/plugins/anchors';
import monospace from '@diplodoc/transform/lib/plugins/monospace';
import imsize from '@diplodoc/transform/lib/plugins/imsize';
import file from '@diplodoc/transform/lib/plugins/file';
import includes from '@diplodoc/transform/lib/plugins/includes';
import tabs from '@diplodoc/transform/lib/plugins/tabs';
import video from '@diplodoc/transform/lib/plugins/video';
import table from '@diplodoc/transform/lib/plugins/table';

import skeletonHandlers from './handlers';
import hooks, {HooksParameters, HooksState} from './hooks';
import {rules} from './rules';
import {mergeHooks} from 'src/hooks';
import {CustomRendererHooks} from '@diplodoc/markdown-it-custom-renderer';
import {CodeRuleState, initState as codeState} from './rules/code-inline';

import {TemplateParameters, generateTemplate} from 'src/xlf/generator';

export type SkeletonRendererState = HooksState & CodeRuleState;

export type RenderParameters = BaseParameters & TemplateParameters & DiplodocParameters;
export type BaseParameters = {
    markdown: string;
    markdownPath?: string;
    skeletonPath?: string;
    hooks?: CustomRendererHooks;
};

export type DiplodocParameters = {
    lang?: string;
};

function render(parameters: RenderParameters) {
    if (!validParameters(parameters)) {
        throw new Error('invalid parameters');
    }

    const {markdown, lang} = parameters;

    const md = new MarkdownIt({html: true}) as HooksParameters['markdownit'];
    const env: MarkdownRendererEnv = {source: markdown.split('\n')};

    const {handlers} = skeletonHandlers.generate();

    const skeletonHooks = hooks.generate({markdownit: md});
    const allHooks: CustomRendererHooks[] = [
        MarkdownRenderer.defaultHooks,
        skeletonHooks.hooks,
    ].concat(parameters.hooks ?? []);
    const mergedHooks = mergeHooks(...allHooks);

    const mdOptions: MarkdownRendererParams<HooksState> = {
        handlers,
        initState: () => ({
            ...skeletonHooks.initState(),
            ...codeState(),
        }),
        rules,
        hooks: mergedHooks,
    };
    const diplodocOptions = {
        lang: lang ?? 'ru',
        path: '',
    };

    // diplodoc plugins
    md.use(meta, diplodocOptions);
    md.use(notes, diplodocOptions);
    md.use(cut, diplodocOptions);
    md.use(sup, diplodocOptions);
    md.use(checkbox, diplodocOptions);
    md.use(anchors, diplodocOptions);
    md.use(monospace, diplodocOptions);
    md.use(imsize, diplodocOptions);
    md.use(file, diplodocOptions);
    md.use(includes, diplodocOptions);
    md.use(tabs, diplodocOptions);
    md.use(video, diplodocOptions);
    md.use(table, diplodocOptions);

    md.use(mdRenderer, mdOptions);

    const skeleton = md.render(markdown, env);
    const state = (md.renderer as MarkdownRenderer<SkeletonRendererState>)
        .state as SkeletonRendererState;
    const xlf = generateTemplate(parameters, state.skeleton.segments);

    return {skeleton, xlf};
}

function validParameters(parameters: RenderParameters) {
    const {markdown, lang} = parameters;

    const markdownCondition = markdown !== undefined;
    const langCondition = lang === undefined || lang === 'ru' || lang === 'en';
    const conditions = [markdownCondition, langCondition];

    return conditions.reduce((a, v) => a && v, true);
}

export {render, validParameters};
export default {render, validParameters};
