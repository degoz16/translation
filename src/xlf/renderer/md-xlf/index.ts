import MarkdownIt from 'markdown-it';
import {
    CustomRendererHooks,
    CustomRendererParams,
    customRenderer,
} from '@diplodoc/markdown-it-custom-renderer';

import hooks from './hooks';
import {mergeHooks} from 'src/utils';
import {rules} from './rules';

import {generateTransUnit} from 'src/xlf/generator';
import {XLFRenderState} from './state';

export {state, XLFRenderState} from './state';

export type XLFRenderParams = DiplodocParams & BaseParams;
export type BaseParams = {
    unitId: number;
    hooks?: CustomRendererHooks;
};

export type DiplodocParams = {
    lang?: string;
};

export function render(tokens: Token[], state: XLFRenderState, parameters: XLFRenderParams) {
    const xlfRenderer = new MarkdownIt({html: true});
    const xlfHooks: {hooks: CustomRendererHooks} = hooks.generate();

    const allHooks: CustomRendererHooks[] = [xlfHooks.hooks].concat(parameters.hooks ?? []);
    const mergedHooks = mergeHooks(...allHooks);
    const xlfOptions: CustomRendererParams<XLFRenderState> = {
        rules: rules(),
        hooks: mergedHooks,
        initState: () => state,
    };

    const env = {
        source: [] as string[],
    };

    xlfRenderer.use(customRenderer, xlfOptions);

    const source = xlfRenderer.renderer.render(tokens, xlfRenderer.options, env);

    return generateTransUnit({
        source,
        id: parameters.unitId,
    });
}
