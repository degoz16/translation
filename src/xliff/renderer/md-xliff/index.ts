import MarkdownIt from 'markdown-it';
import {
    CustomRendererHooks,
    customRenderer,
} from '@diplodoc/markdown-it-custom-renderer';
import {token} from 'src/utils';

import {hooks} from './hooks';
import {rules} from './rules';

import {generateTransUnit} from 'src/xliff/generator';
import {XLFRenderState} from './state';

export {state, XLFRenderState} from './state';

export type XLFRenderParams = DiplodocParams & BaseParams;
export type BaseParams = {
    unitId: number;
    hooks?: CustomRendererHooks;
};

export type DiplodocParams = {
    lang?: string;
};

export function render(tokens: Token[], state: XLFRenderState, parameters: XLFRenderParams) {
    const xliffRenderer = new MarkdownIt({html: true});

    xliffRenderer.use(customRenderer, {
        rules,
        hooks,
        initState: () => state,
    });

    const source = xliffRenderer.renderer.render([token('inline', {
        children: groupUselessTokens(tokens)
    })], xliffRenderer.options, {
        source: [] as string[],
    });

    return generateTransUnit({
        source,
        id: parameters.unitId,
    });
}

function groupUselessTokens(tokens: Token[]) {
    const map: Record<string, Token[]> = {};
    const result = [];

    for (const part of tokens) {
        if (!part.content) {
            const [name, type] = part.type.split('_');
            if (type === 'open') {
                map[part.type] = map[part.type] || [];
                map[part.type].push(part);
            } else if (type === 'close') {
                map[name + '_open'] = map[name + '_open'] || [];
                const opener = map[name + '_open'].pop();

                if (opener) {
                    opener.g = part;
                    part.g = opener;
                }
            }
        }

        result.push(part);
    }

    return result;
}