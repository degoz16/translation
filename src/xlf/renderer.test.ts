import {render, RenderParameters} from './renderer';

describe('smoke', () => {
    test('it works', () => {
        const parameters = {
            markdown: '',
            source: {
                language: 'ru',
                locale: 'RU' as const,
            },
            target: {
                language: 'en',
                locale: 'US' as const,
            },
            skeletonPath: 'file.skl.md',
            markdownPath: 'file.md',
        };

        render(parameters);
    });
});

describe('validates parameters', () => {
    it('works with valid parameters', () => {
        const parameters = {
            markdown: '',
            source: {
                language: 'ru',
                locale: 'RU' as const,
            },
            target: {
                language: 'en',
                locale: 'US' as const,
            },
            skeletonPath: 'file.skl.md',
            markdownPath: 'file.md',
        };

        render(parameters);
    });

    it('throws on invalid parameters', () => {
        const parameters = {
            source: {
                language: 'ru',
                locale: 'RU' as const,
            },
            target: {
                language: 'en',
                locale: 'US' as const,
            },
            skeletonPath: 'file.skl.md',
            markdownPath: 'file.md',
        } as RenderParameters;

        expect(() => render(parameters)).toThrow();
    });
});

describe('xlf rendering', () => {
    it('renders empty xlf document for empty markdown', () => {
        const parameters = {
            markdown: '',
            source: {
                language: 'ru',
                locale: 'RU' as const,
            },
            target: {
                language: 'en',
                locale: 'US' as const,
            },
            skeletonPath: 'file.skl.md',
            markdownPath: 'file.md',
        };

        const generated = render(parameters);

        expect(generated).toMatchSnapshot();
    });
});
