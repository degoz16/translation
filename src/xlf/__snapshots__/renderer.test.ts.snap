// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`xlf rendering renders empty xlf document for empty markdown 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<xliff xmlns=\\"urn:oasis:names:tc:xliff:document:1.2\\" version=\\"1.2\\">
  <file original=\\"file.md\\" source-language=\\"ru-RU\\" target-language=\\"en-US\\" datatype=\\"markdown\\">
    <header>
      <skeleton>
        <external-file href=\\"file.skl.md\\"></external-file>
      </skeleton>
    </header>
    <body>
    </body>
  </file>
</xliff>
"
`;
